orca/producer.go

```diff
@@ -0,0 +1,219 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package orca
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/orca/internal"
+	"google.golang.org/grpc/status"
+
+	v3orcapb "github.com/cncf/xds/go/xds/data/orca/v3"
+	v3orcaservicegrpc "github.com/cncf/xds/go/xds/service/orca/v3"
+	v3orcaservicepb "github.com/cncf/xds/go/xds/service/orca/v3"
+	"google.golang.org/protobuf/types/known/durationpb"
+)
+
+type producerBuilder struct{}
+
+// Build constructs and returns a producer and its cleanup function
+func (*producerBuilder) Build(cci interface{}) (balancer.Producer, func()) {
+	ctx, cancel := context.WithCancel(context.Background())
+	p := &producer{
+		client:    v3orcaservicegrpc.NewOpenRcaServiceClient(cci.(grpc.ClientConnInterface)),
+		closed:    grpcsync.NewEvent(),
+		intervals: make(map[time.Duration]int),
+		listeners: make(map[OOBListener]struct{}),
+		backoff:   internal.DefaultBackoffFunc,
+	}
+	go p.run(ctx)
+	return p, func() {
+		cancel()
+		<-p.closed.Done() // Block until stream stopped.
+	}
+}
+
+var producerBuilderSingleton = &producerBuilder{}
+
+// OOBListener is used to receive out-of-band load reports as they arrive.
+type OOBListener interface {
+	// OnLoadReport is called when a load report is received.
+	OnLoadReport(*v3orcapb.OrcaLoadReport)
+}
+
+// OOBListenerOptions contains options to control how an OOBListener is called.
+type OOBListenerOptions struct {
+	// How often to request the server to provide a load report.  May be
+	// provided less frequently if the server requires a longer interval, or
+	// may be provided more frequently if another subscriber requests a shorter
+	// interval.
+	ReportInterval time.Duration
+}
+
+// RegisterOOBListener registers an out-of-band load report listener on sc.
+// Any OOBListener may only be registered once per subchannel at a time.  The
+// returned stop function must be called when no longer needed.
+func RegisterOOBListener(sc balancer.SubConn, l OOBListener, opts OOBListenerOptions) (stop func()) {
+	pr, close := sc.GetOrBuildProducer(producerBuilderSingleton)
+	p := pr.(*producer)
+	p.registerListener(l, opts.ReportInterval)
+
+	// TODO: When we can register for SubConn state updates, don't call run()
+	// until READY and automatically call stop() on SHUTDOWN.
+
+	// If stop is called multiple times, prevent it from having any effect on
+	// subsequent calls.
+	var once sync.Once
+	return func() {
+		once.Do(func() {
+			p.unregisterListener(l, opts.ReportInterval)
+			close()
+		})
+	}
+}
+
+type producer struct {
+	client v3orcaservicegrpc.OpenRcaServiceClient
+
+	ctx     context.Context
+	cancel  func()
+	closed  *grpcsync.Event // fired when closure completes
+	backoff func(int) time.Duration
```

Comment:
Reword these three lines: "To avoid hammering a server that may be experiencing problems, the client will use exponential backoff between attempts. When the client receives a message from the server on a given call, the backoff state is reset, so the next attempt will occur immediately, but any subsequent attempt will be subject to exponential backoff." from A51. Maybe something like // backoff is used to avoid overloading servers not behaving properly. Reset when receives a message from server on a call.

Reply:
Done.  PTAL.

Reply:
Nice, thanks.